File: src/formatters.ts
import type { Selection, TextDocument } from 'vscode';
import type { FormatConfig } from './config';

function prefix(cfg: FormatConfig): string {
  return cfg.atPrefix ? '@' : '';
}

function separator(cfg: FormatConfig): string {
  return cfg.rangeSeparator === 'colon' ? ':' : '-';
}

function lineSuffix(selection: Selection, cfg: FormatConfig): string {
  const startLine = selection.start.line + 1;
  const endLine = selection.end.line + 1;
  if (startLine === endLine) {
    return `:${startLine}`;
  }
  return `:${startLine}${separator(cfg)}${endLine}`;
}

export function formatPathLine(relPath: string, selection: Selection, cfg: FormatConfig): string {
  return `${prefix(cfg)}${relPath}${lineSuffix(selection, cfg)}`;
}

export function formatPath(relPath: string, cfg: FormatConfig): string {
  return `${prefix(cfg)}${relPath}`;
}

export function formatFullPath(absPath: string, cfg: FormatConfig): string {
  return `${prefix(cfg)}${absPath}`;
}

function fenceFor(code: string): string {
  // CommonMark: a fenced block's closing delimiter must be at least as long as
  // the opening, and any backtick run inside the body must be shorter than the
  // fence. Pick fence length = max(3, longest backtick run + 1).
  let maxRun = 0;
  let currentRun = 0;
  for (const ch of code) {
    if (ch === '`') {
      currentRun++;
      if (currentRun > maxRun) {
        maxRun = currentRun;
      }
    } else {
      currentRun = 0;
    }
  }
  return '`'.repeat(Math.max(3, maxRun + 1));
}

export function formatMarkdown(
  relPath: string,
  selection: Selection,
  document: TextDocument,
  cfg: FormatConfig,
): string {
  const pathLabel = formatPathLine(relPath, selection, cfg);
  const language = document.languageId ?? '';

  let code: string;
  if (selection.isEmpty) {
    code = document.lineAt(selection.start.line).text;
  } else {
    code = document.getText(selection);
  }

  const fence = fenceFor(code);
  return `${pathLabel}\n${fence}${language}\n${code}\n${fence}`;
}

File: nvim/lua/copy-path-line/formatters.lua
local M = {}

local function prefix(cfg)
  return cfg.at_prefix and '@' or ''
end

local function range_sep(cfg)
  return cfg.range_separator == 'colon' and ':' or '-'
end

local function line_suffix(range, cfg)
  if range.start_line == range.end_line then
    return ':' .. range.start_line
  end
  return ':' .. range.start_line .. range_sep(cfg) .. range.end_line
end

function M.format_path_line(rel_path, range, cfg)
  return prefix(cfg) .. rel_path .. line_suffix(range, cfg)
end

function M.format_path(rel_path, cfg)
  return prefix(cfg) .. rel_path
end

function M.format_full_path(abs_path, cfg)
  return prefix(cfg) .. abs_path
end

local function fence_for(text)
  local max_run = 0
  local current = 0
  for i = 1, #text do
    if text:sub(i, i) == '`' then
      current = current + 1
      if current > max_run then
        max_run = current
      end
    else
      current = 0
    end
  end
  local n = math.max(3, max_run + 1)
  return string.rep('`', n)
end

function M.format_markdown(rel_path, range, lines, language, cfg)
  local label = M.format_path_line(rel_path, range, cfg)
  local code = table.concat(lines, '\n')
  local fence = fence_for(code)
  return label .. '\n' .. fence .. (language or '') .. '\n' .. code .. '\n' .. fence
end

return M
